entity-framework/core/querying/complex-query-operators.md

```diff
@@ -0,0 +1,128 @@
+
+### Collection selector doesn't reference outer
+
+When collection selector isn't referencing anything from the outer source, the result is a cartesian product of both data sources. It translates to `CROSS JOIN` in relational databases.
+
+[!code-csharp[Main](../../../samples/core/Querying/ComplexQuery/Sample.cs#SelectManyConvertedToCrossJoin)]
+
+```SQL
+SELECT [b].[BlogId], [b].[OwnerId], [b].[Rating], [b].[Url], [p].[PostId], [p].[AuthorId], [p].[BlogId], [p].[Content], [p].[Rating], [p].[Title]
+FROM [Blogs] AS [b]
+CROSS JOIN [Posts] AS [p]
+```
+
+### Collection selector references outer in where clause
+
+When collection selector has a where predicate, which is referencing the outer element, then EF Core translates it to join on database and uses the predicate as join condition. Normally it arises when using collection navigation on outer source as collection selector. If the collection is empty for outer element, then it wouldn't generate any result for that outer element. But if `DefaultIfEmpty` is applied on collection selector then outer element will be connected with default value of inner element. Because of this distinction, this kind of queries translates to `INNER JOIN` in the absence of `DefaultIfEmpty` and `LEFT JOIN` when `DefaultIfEmpty` is applied.
+
+[!code-csharp[Main](../../../samples/core/Querying/ComplexQuery/Sample.cs#SelectManyConvertedToJoin)]
+
+```SQL
+SELECT [b].[BlogId], [b].[OwnerId], [b].[Rating], [b].[Url], [p].[PostId], [p].[AuthorId], [p].[BlogId], [p].[Content], [p].[Rating], [p].[Title]
+FROM [Blogs] AS [b]
+INNER JOIN [Posts] AS [p] ON [b].[BlogId] = [p].[BlogId]
+
+SELECT [b].[BlogId], [b].[OwnerId], [b].[Rating], [b].[Url], [p].[PostId], [p].[AuthorId], [p].[BlogId], [p].[Content], [p].[Rating], [p].[Title]
+FROM [Blogs] AS [b]
+LEFT JOIN [Posts] AS [p] ON [b].[BlogId] = [p].[BlogId]
+```
+
+### Collection selector references outer in non-predicate
+
+When collection selector has reference to outer, which isn't in predicate (as the above case), it can't be converted to join. That's why we need to evaluate collection selector for each outer element. It translates to `APPLY` operation in relational databases. If the collection is empty for outer element, then it wouldn't generate any result for that outer element. But if `DefaultIfEmpty` is applied on collection selector then outer element will be connected with default value of inner element. Because of this distinction, this kind of queries translates to `CROSS APPLY` in the absence of `DefaultIfEmpty` and `OUTER APPLY` when `DefaultIfEmpty` is applied. Sqlite doesn't support `APPLY` operator.
```

Comment:
This is quite SQL Server-specific.

```suggestion
When the collection selector references the outer element, but from a non-predicate operator (as was the  case above), the query can't be translated to a regular database join, since we need to evaluate collection selector for each outer element. However, the query does translate to an `APPLY` or a lateral join operation in relational databases.

As before, if the collection is empty for the outer element, then it wouldn't generate any result for that outer element. But if `DefaultIfEmpty` is applied on collection selector then outer element will be connected with the default value of inner element type. Because of this distinction, on SQL Server this kind of query translates to `CROSS APPLY` in the absence of `DefaultIfEmpty` and `OUTER APPLY` when `DefaultIfEmpty` is applied. 

Sqlite doesn't support `APPLY` operator, and therefore this kind of query cannot be translated when targeting it.
```

Reply:
Lateral join and apply operations are different. It is fine the way it is. It is Oracle-specific.

Reply:
They're not really that different. PostgreSQL implements the APPLY behavior with a LATERAL JOIN, there's just an additional ON TRUE there.

Reply:
Again Lateral joins are apply operations are different. Postgre fits apply operation on lateral join with additional condition. I have added statement that query may not translate always from EF docs perspective, we generate Apply operation and that is accurate doc content.

Reply:
The fact is that what this section describes (Collection selector references outer in non-predicate) is translated in PostgreSQL via lateral join, which is part of the SQL standard; PostgreSQL users reading this will have no idea about APPLY. All I'm asking is to mention lateral join alongside APPLY so that they understand, because that's what is used for translation.

Reply:
That is misleading. Relational generates Apply operation. EF Core relational layer does not generate lateral joins. Writing lateral join easily confuses user asking, oh why did not translate this query via lateral join. Sorry but we don't support lateral join.

Reply:
I think I'll let this PR go now.